File: contests/lc/611.ValidTriangleNumber/valid-triangle-number.go
package validtrianglenumber

import "sort"

// triangleNumber counts index triplets (i, j, k) whose values can form
// a non-degenerate triangle. After sorting we iterate the first two sides
// and binary search the longest valid third side boundary in the suffix.
func triangleNumber(nums []int) int {
	sort.Ints(nums)
	ans := 0
	for i, n := 0, len(nums); i < n-2; i++ {
		for j := i + 1; j < n-1; j++ {
			left, right := j+1, n
			for left < right {
				mid := (left + right) >> 1
				if nums[mid] >= nums[i]+nums[j] {
					right = mid
				} else {
					left = mid + 1
				}
			}
			ans += left - j - 1
		}
	}
	return ans
}

File: contests/lc/611.ValidTriangleNumber/valid-triangle-number.py
from bisect import bisect_left
from typing import List

class Solution:
    def triangleNumber(self, nums: List[int]) -> int:
        nums.sort()
        ans, n = 0, len(nums)
        for i in range(n - 2):
            for j in range(i + 1, n - 1):
                k = bisect_left(nums, nums[i] + nums[j], lo=j + 1) - 1
                ans += k - j
        return ans
    
def main():
    s = Solution()
    print(s.triangleNumber([2,2,3,4]))  # Output: 3
    print(s.triangleNumber([4,2,3,4]))  # Output: 4

if __name__ == "__main__":
    main()